R/aggregateSolute.R

```diff
@@ -198,60 +200,22 @@ aggregateSolute <- function(
-    # suggested by Newey and West (1987); other weights were suggested in papers 
-    # subsequent to that one. If you do use such a weighting function, you should
-    # choose L wisely; it should probably be constant across all aggregation
-    # periods rather than changing from period to period, and it should probably
-    # depend on the temporal resolution of your predictions. The weight function
-    # is sometimes also called a kernel function.
-    #   weight.fun=function(lag.num) { rep(1, length(lag.num)) }
-    #   acfvec <- acfvec * weight.fun(0:num.lags)
-    
-    # First compute the correlations among observation errors, based on
-    # assumptions or estimates embodied in cormat.function
-    cor_matrix <- cormat.function(dates)
-    
-    # The covariance matrix is the element-wise product of cor_matrix and 
-    # varvar_matrix, where varvar_matrix[i,j] is se.preds[i]*se.preds[j]. 
-    # Because what we really want is the sum of all terms in the covariance 
-    # matrix, and because varvar_matrix could be huge, we don't actually ever 
-    # create varvar_matrix. Instead we use sparse matrix multiplication to get 
-    # us first to another sparse matrix (half_cov_matrix) and then to a column 
-    # vector (sums by row of products), and then we take the sum of that vector
-    # straight away.
-    half_cov_matrix <- cor_matrix*se.preds
-    cov_matrix_sum <- sum(half_cov_matrix %*% se.preds)
-    
-    # At this point (between creating cov_matrix and computing the summation,
-    # actually), the AMLE algorithm in rloadest adds an additional term for
-    # covariance arising from coefficient uncertainty - see src/TAC_LOAD.f - but
-    # for other model types this additional covariance is captured in se.pred
-    # and the chosen cormat.function already.
-    
-    # Return the calculated SE
-    sqrt(cov_matrix_sum)/length(se.preds)
-  }
-  agg_preds <- v(data.frame(preds, se.preds, dates, aggregate_by)) %>%
-    group_by_(.dots=as.list(agg.by)) %>%
-    dplyr::summarise(Value=mean(preds), SE=if(se.agg | ci.agg) SEofSum(dates, se.preds) else NA) %>%
-    as.data.frame()
+  #get threshold for number of measurements
+  n_threshold <- switch(agg.by,
+                      "month"=28*complete.threshold,
+                      "water_year"=365*complete.threshold, 
+                      "calendar_year"=365*complete.threshold,
+                      0)
+  
+  #do the actual stats on grouped df
+  preds_grp <- group_by_(v(data.frame(preds, se.preds, dates, aggregate_by)), 
+                          .dots=as.list(agg.by)) 
+  #drop data for incomplete units with filter
+  agg_preds <- as.data.frame(summarise(filter(preds_grp, n() > n_threshold), 
+                                      Value=mean(preds), 
+                                      SE=if(se.agg | ci.agg) SEofSum(dates, se.preds, cormat.function) else NA,
+                                      n = if(agg.by != "unit") n() else NULL)) 
```

Comment:
as you're working through this, think you could formalize a few tests into testthat tests? `test-33-aggregateSolute` is ripe for an update. 

(that file is currently excluded from the automated tests because 1+ of the loadReg calls is segfaulting on travis. we might have to leave it that way for now, but i'd be comforted to know that you'd introduced a few tests that we can run locally now and add back to the automated testing eventually.)